### PR TITLE
[pytorch][codegen] migrate gen_variable_type to new data model

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49735 [pytorch][codegen] migrate gen_variable_type to new data model**

This is the final wave of autograd codegen data model migration.

After this PR:
- autograd codegen no longer depends on Declarations.yaml;
- autograd codegen sources are fully type annotated and pass mypy-strict check;

To avoid potential merge conflicts with other pending PRs, some structural
changes are intentionally avoided, e.g. didn't move inner methods out, didn't
change all inner methods to avoid reading outer function's variables, and etc.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Confirmed clean mypy-strict run:
```
mypy --config mypy-strict.ini
```

Differential Revision: [D25678879](https://our.internmc.facebook.com/intern/diff/D25678879)

### CODE REVIEW

**Comment by ljk53** [comment]
@albanD I also have it rebased on top of your PR: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa

The merge conflicts seem to be manageable - so I think this could go before or after your PR, let me know your plan :)

File: tools/autograd/gen_variable_type.py
Comment (albanD) [inline tools/autograd/gen_variable_type.py:423]: nit: `argumenttype_type` is a bit surprising in terms of naming convention :)

File: tools/autograd/gen_variable_type.py
Comment (albanD) [inline tools/autograd/gen_variable_type.py:575]: Well I guess an output can only be differentiable if it is tensor like. So, I would expect this check to not do much?

File: tools/autograd/gen_variable_type.py
Comment (albanD) [inline tools/autograd/gen_variable_type.py:722]: Out of curiosity, is there a difference between `"` and `'` in python? Or is this just for consistency?

File: tools/autograd/gen_variable_type.py
Comment (albanD) [inline tools/autograd/gen_variable_type.py:756]: We cannot check `is_tensor_like()` here?

File: tools/autograd/gen_variable_type.py
Comment (albanD) [inline tools/autograd/gen_variable_type.py:932]: Given the replacements above, it should not. Otherwise we should add an argument to disable them for most use cases here.

**Review by albanD — COMMENTED** [review]
That looks quite good!
Small comments below.

**Review by ezyang — APPROVED** [review]
Approving to unblock

File: tools/autograd/gen_autograd.py
Comment (bhosmer) [inline tools/autograd/gen_autograd.py:109]: It's probably worth pulling the inline imports (here and line 106, at least) out to the top if you're back in here before landing.

File: tools/autograd/gen_variable_type.py
Comment (bhosmer) [inline tools/autograd/gen_variable_type.py:281]: note to self: understand why this needs to be optional

File: tools/autograd/gen_variable_type.py
Comment (bhosmer) [inline tools/autograd/gen_variable_type.py:971]: I could definitely be missing something obvious, but when might a `None` creep in here (necessitating its optionality)?

File: tools/jit/gen_unboxing_wrappers.py
Comment (bhosmer) [reply tools/jit/gen_unboxing_wrappers.py:306]: @albanD JFYI after this PR the mainline autograd codegen no longer uses this at all, and after Sebastian's c10-full stuff it should be gone completely (along with `gen_unboxing_wrappers.py`)

**Review by bhosmer — APPROVED** [review]
Hey, looks good! Couple of very minor things noted inline, but not worth holding up landing, only if you're back in for some other reason.

(Though I don't know enough to know if Alban's question on `DifferentiableOutput` implies an issue.)

File: tools/jit/gen_unboxing_wrappers.py
Comment (ljk53) [reply tools/jit/gen_unboxing_wrappers.py:290]: This entire file is going to be deprecated and the modifications are all simple copy-paste.
I moved old flavor code from gen_autograd to here so all autograd code can pass mypy-strict check.

File: tools/autograd/gen_autograd.py
Comment (ezyang) [reply tools/autograd/gen_autograd.py]: fwiw, the codegen is still simple enough and free of BC concerns that we should be able to come up with some organization that doesn't have circular problems. But I agree, it doesn't matter for codegen, inline import or whatever you want to fix the problem is fine.

File: tools/autograd/gen_variable_type.py
Comment (bhosmer) [reply tools/autograd/gen_variable_type.py:311]: Ah sure - might be worth pasting your answer in here as a comment, if you do another edit to this file.

File: tools/codegen/api/autograd.py
Comment (albanD) [reply tools/codegen/api/autograd.py:108]: My question is just about the comment.
But in here, when there is a `mul_` in native_functions but only `mul` defined in derivatives, we re-use the formula and do things like replacing "result" by "self".
It is just to try and capture in the comment the exact state that is saved here: before or after the adaptation based on the exact variant we're working with.

File: tools/codegen/api/types.py
Comment (albanD) [reply tools/codegen/api/types.py:43]: I just find the structure of the function hard to follow. Maybe an `else` statement would make it clearer?
But this is just a style thing, I don't think this the code is wrong.
But something like this is more natural to me:
```python
no_ref_type = self.elem.cpp_type()
if strip_ref:
  return no_ref_type
else:
  return f'const {no_ref_type} &'
```

File: tools/jit/gen_unboxing_wrappers.py
Comment (albanD) [reply tools/jit/gen_unboxing_wrappers.py:290]: This was not related to mypy actually. I meant in the generated string so that the c++ codegened code is easier to read

File: tools/autograd/gen_autograd.py
Comment (bhosmer) [reply tools/autograd/gen_autograd.py]: > Curious - what's the problem with inline import? I remember you mentioned we discovered some performance issue with inline import, but why it matters for codegen?

Oh definitely not a perf concern here (or most places), just hygiene.

File: tools/autograd/gen_variable_type.py
Comment (ljk53) [inline tools/autograd/gen_variable_type.py:1010]: oops - I think I missed this check while solving merge conflicts... sorry!

I had a revision with your PR fully patched which contains this check: https://github.com/ljk53/pytorch/commit/313f5cc50b4c2eaa50128bd08d2efe19c998effa#diff-f6553cec68caeaea36f6c8b14ff76a6d39dfd774e0ea9ef2f76e8d81fd9af5dfR1096-R1105

@albanD I can send out a PR to add this check back, or you can help fix it if you are about to land your final codegen PR - please let me know :)

Comment (albanD) [reply tools/autograd/gen_variable_type.py:1010]: I can add that back sure. Thanks!

**Pull request closed without merging**
